GDScript: Add module description in markdown (#81345)
godotengine/godot: modules/gdscript/README.md

Author: This makes sense somewhere in the contributing to the engine part, right?

Author: Oh, I understand now! So something like "If you're looking to learn more about GDScript, please see here", and that links to https://docs.godotengine.org/en/latest/tutorials/scripting/gdscript/index.html?

Reviewer: Maybe something like "You can learn more about GDScript in the [documentation](https://docs.godotengine.org/en/latest/tutorials/scripting/gdscript/index.html). It describes the syntax and user facing systems and concepts, and can be used as a reference for what user expectations are."

## General design

GDScript is:

1. A [gradually typed](https://en.wikipedia.org/wiki/Gradual_typing) language. Type hints are optional and help with static analysis and performance. However, typed code must easily interoperate with untyped code.
2. A tightly designed language. Features are added because they are _needed_, and not because they can be added or are interesting to develop.
3. Primarily an interpreted scripting language: it is compiled to GDScript byte code and interpreted in a GDScript virtual machine. It is meant to be easy to use and develop gameplay in. It is not meant for CPU-intensive algorithms or data processing, and is not optimized for it. For that, [C#](https://docs.godotengine.org/en/stable/tutorials/scripting/c_sharp/c_sharp_basics.html) or [GDExtension](https://docs.godotengine.org/en/stable/tutorials/scripting/gdextension/what_is_gdextension.html) may be used.

## Integration into Godot

> GDScript is integrated into godot as a module. Since modules are optional, this means that Godot may be built without GDScript and work perfectly fine without it!

Reviewer 2: ```suggestion
GDScript is integrated into Godot as a module. Since modules are optional, this means that Godot may be built without GDScript and work perfectly fine without it!
```

Author: Oh, nice catch! Thanks, and fixed :)

> The GDScript module interfaces with Godot's codebase by inheriting from Godot's scripting-related classes. New languages inherit from [`ScriptLanguage`](/core/object/script_language.h), and are registered in Godot's [`ScriptServer`](/core/object/script_language.h). Scripts, referring to a file containing code, inherit from `Script`. Instances of that script, which are used at runtime when actually executing the code, inherit from [`ScriptInstance`](/core/object/script_instance.h).

Author: Adding these changes, except for mentioning the "resource" type. It's a concept they might not know/understand just yet and which isn't super relevant for immediately understanding GDScript - might be a case of less is more here? :)

> To access Godot's internal classes, GDScript uses [`ClassDB`](/core/object/class_db.h). `ClassDB` is where Godot registers classes, methods and properties that it wants exposed to its scripting system. This is how GDScript understands what a `Vector3` is, that `print()` exists, or that `Node` has a `get_parent()` method.

Reviewer 2: I'm not sure the last sentence is accurate. Isn't ClassDB only for Object-derived classes?
I believe both Variant types and GDScript methods are registered via a different mechanism.

Should be confirmed, I'm no expert.

### Loading source code

This mostly happens by calling `GDScript::load_source_code()` on a `GDScript` object. Parsing only requires a `String`, so it is entirely possible to parse a script without a `GDScript` object!

### Tokenizing (see [`GDScriptTokenizer`](gdscript_tokenizer.h))

Tokenizing is the process of converting the source code `String` into a sequence of tokens, which represent language constructs (such as `for` or `if`), identifiers, literals, etc. This happens almost exclusively during the parsing process, which asks for the next token in order to make sense of the source code. The tokenizer is only used outside of the parsing process in very rare exceptions.

### Parsing (see [`GDScriptParser`](gdscript_parser.h))

The parser takes a sequence of tokens and builds [the abstract syntax tree (AST)](https://en.wikipedia.org/wiki/Abstract_syntax_tree) of the GDScript program. The AST is used in the analyzing and compilation steps, and the source code `String` and sequence of tokens are discarded. This is the process that finds any syntax errors in a GDScript program and reports them to the user. 

> The parser class also defines all the possible nodes of the AST, e.g. `GDScriptParser::ClassNode` for a node representing a GDScript class, or `GDScriptParser::FunctionNode` for a function. It further defines typechecking datastructures like `GDScriptParser::Datatype`.

Reviewer 2: Is it worth clarifying that "node" in this context is unrelated to the Node class?

> The parser class also defines all the possible nodes of the AST, e.g. `GDScriptParser::ClassNode` for a node representing a GDScript class, or `GDScriptParser::FunctionNode` for a function. It further defines typechecking datastructures like `GDScriptParser::Datatype`.

Reviewer 2: ```suggestion
The parser class also defines all the possible nodes of the AST, e.g. `GDScriptParser::ClassNode` for a node representing a GDScript class, or `GDScriptParser::FunctionNode` for a function. It further defines typechecking data structures like `GDScriptParser::Datatype`.
```

The analyzer is where a lot of the "magic" happens! It defines what constitutes proper code that can actually be compiled or run.

This is also where dependencies are introduced. If a class `A` extends a class `B` or contains a member with type `B` from some other script file, then the analyzer will attempt to load that second script. If `B` contains references to `A`, then a _cyclic_ dependency is introduced.

#### Cyclic dependencies and member resolution

Cyclic dependencies from inheritance (`A extends B, B extends A`) are not supported in any programming language. Other cyclic dependencies are supported, such as `A extends B` and `B` contains, or preloads, members of type `A`. This is possible because of two features of the analyzer:

1. Class interface resolution: to know what is available from one class, you don't need to analyze it in its entirety. You may merely analyze its "interface", meaning its name and its members (and their types!), as well as name and members of its base case, recursively. Details can be found in `GDScriptAnalyzer::resolve_class_interface()`.

> 2. Out of order member resolution: the analyzer may not need an entire class interface to be resolved in order to figure out a specific type. For example, if class `A` contains code that references `B.is_alive`, then the analyzer doesn't need to immediately resolve `B`'s interface. It may simply check whether `is_alive` exists in `B`. This pattern happens often in the analyzer, and can be easily identified by the declaration of a temporary datatype with `GDScriptParser::DataType resolving_datatype;`, followed by `resolving_datatype.kind = GDScriptParser::DataType::RESOLVING;`. If the analyzer attempts to resolve a member that is already resolving, then a cyclic dependency problem has been found.

Reviewer 2: Up for discussion, I can see both being accepted. The class is `DataType` so two words, but the local itself is `resolving_datatype`.
```suggestion
2. Out of order member resolution: the analyzer may not need an entire class interface to be resolved in order to figure out a specific type. For example, if class `A` contains code that references `B.is_alive`, then the analyzer doesn't need to immediately resolve `B`'s interface. It may simply check whether `is_alive` exists in `B`. This pattern happens often in the analyzer, and can be easily identified by the declaration of a temporary data type with `GDScriptParser::DataType resolving_datatype;`, followed by `resolving_datatype.kind = GDScriptParser::DataType::RESOLVING;`. If the analyzer attempts to resolve a member that is already resolving, then a cyclic dependency problem has been found.
```

Author: If it's OK, I think I prefer the way it is right now since like you mention it is `resolving_datatype` and not `resolving_data_type`.  This way we don't have to worry about writing data type vs datatype depending on context!

### Compiling (see [`GDScriptCompiler`](gdscript_compiler.h))

Compiling is the final step in making a GDScript executable in the [virtual machine](gdscript_vm.h) (VM). The compiler takes a `GDScript` object and an AST, and uses another class, [`GDScriptByteCodeGenerator`](gdscript_byte_codegen.h), to generate VM instructions corresponding to the class. In doing this, it generates the objects that the VM understands how to run, like [`GDScriptFunction`](gdscript_function.h), and completes a few extra tasks needed for compilation, such as populating class member information.

Importantly, the compilation process of a class, specifically `GDScriptCompiler::_compile_class()` _cannot_ depend on the compilation process of another class, since that will result in cyclic dependency problems. Supported cyclic dependencies must have been fully resolved during the analyze step. It is OK for supporting methods, such as `GDScriptCompiler::_populate_class_methods()` to be called on other classes as long as those methods rely exclusively on information from previous steps.

## Loading scripts

> GDScripts can be loaded in a couple of different ways. The main method that the Godot itself uses to load scripts is through the `ResourceLoader` singleton. In this way, GDScripts are resources like any others, and `ResourceLoader::load()` will route to `ResourceFormatLoaderGDScript::load()` found in `gdscript.h/cpp`(gdscript.h). This generates a GDScript object which is compiled and ready to use.

Reviewer 2: ```suggestion
GDScripts can be loaded in a couple of different ways. The main method that Godot itself uses to load scripts is through the `ResourceLoader` singleton. In this way, GDScripts are resources like any others, and `ResourceLoader::load()` will route to `ResourceFormatLoaderGDScript::load()` found in [`gdscript.h/cpp`](gdscript.h). This generates a GDScript object which is compiled and ready to use.
```

> The other method is to manually load the source code, then pass it to a parser, then to an analyzer and then to a compiler. This is what happens behind the scenes of the previous approach, minus some caching of GDScripts. It is used in the [GDScript test runner infrastructure](tests/gdscript_test_runner.h`).

Reviewer 2: ```suggestion
The other method is to manually load the source code, then pass it to a parser, then to an analyzer and then to a compiler. This is what happens behind the scenes of the previous approach, minus some caching of GDScripts. It is used in the [GDScript test runner infrastructure](tests/gdscript_test_runner.h).
```

## Special types of scripts

Certain types of GDScripts behave slightly differently. For example, autoloads are loaded with `ResourceLoader::load()` during `Main::start()`, very soon after Godot is launched. Many systems aren't initialized at that time, so error reporting is often significantly reduced and may not even show up in the editor.

Tool scripts, declared with the `@tool` annotation on a GDScript file, run in the editor itself as opposed to just when the game is launched. This leads to a significant increase in complexity, as many things that can be changed in the editor may affect a currently executing tool script.

## Other

There are many other classes in the GDScript module. Here is a brief overview of some of them:

- Declaration of GDScript warnings in [`GDScriptWarning`](gdscript_warning.h).
- [`GDScriptFunction`](gdscript_function.h), which represents an executable GDScript function. The relevant file contains both static as well as runtime information.
- The [virtual machine](gdscript_vm.cpp) is essentially defined as calling `GDScriptFunction::call()`.

> - Editor related functions can be found in parts of `GDScriptLanguage`, originally declared in [`gdscript.h`](gdscript.h) but defined in [`gdscript_editor.cpp`](gdscript_editor.cpp). Code highlighting can be found in [`GDScriptSyntaxHighlighter`](editor/gdscript_highlighter.h).

Reviewer 2: ```suggestion
- Editor-related functions can be found in parts of `GDScriptLanguage`, originally declared in [`gdscript.h`](gdscript.h) but defined in [`gdscript_editor.cpp`](gdscript_editor.cpp). Code highlighting can be found in [`GDScriptSyntaxHighlighter`](editor/gdscript_highlighter.h).
```

> 

Reviewer: Is this empty line accidental?

Author: Nope, really wasn't! :)

Reviewer: You mean we need an empty line for some reason?

Author: Oh my goodness, my bad! I meant it really wasn't _intentional_, not _accidental_ 😅

# Basic GDScript module architecture
This provides some basic information in how GDScript is implemented and integrates with the rest of the engine. You can learn more about GDScript in the [documentation](https://docs.godotengine.org/en/latest/tutorials/scripting/gdscript/index.html). It describes the syntax and user facing systems and concepts, and can be used as a reference for what user expectations are.

## General design

GDScript is:

1. A dynamically typed language. Type hints help with static analysis and performance, but typed code must easily interoperate with untyped code.
2. A tightly designed language. Features are added because they are _needed_, and not because they can be added or are interesting to develop.

> 3. Primarily an interpreted scripting language. It is meant to be easy to use and develop gameplay in. It is not meant for CPU-intensive algorithms or data processing, and is not optimized for it. For that, C# or GDExtensions may be used.

Reviewer 3: ```suggestion
3. Primarily an interpreted scripting language. It is meant to be easy to use and develop gameplay in. It is not meant for CPU-intensive algorithms or data processing, and is not optimized for it. For that, C# or [GDExtension](https://docs.godotengine.org/en/stable/tutorials/scripting/gdextension/what_is_gdextension.html) may be used.
```

## Compilation
Scripts can be at different stages of compilation, but the process generally consists of a linear series of steps: tokenizing, parsing, analyzing, and finally compiling.

> The main class of the GDScript module is the [`GDScript`](gdscript.h) class, which represents a GDScript file, and contains information regarding all that that file contains, such as main class and references to subclasses, members, signals, implicit methods like initializers, etc. This is the main class that the compilation step deals with.

Reviewer 3: ```suggestion
The main class of the GDScript module is the [`GDScript`](gdscript.h) class, which usually represents a GDScript file, and contains information regarding all that that file contains, such as main class and references to subclasses, members, signals, implicit methods like initializers, etc. This is the main class that the compilation step deals with.

```note
Each subclass are themselves `GDScript` instances.
```
```

> # Basic GDScript module architecture

Reviewer 4: Inconsistent blank lines in the file. I suggest using 2 blank lines between sections and 1 blank line to separate the body from the header.

```md
## Header

Text.

## Next header

Text.
```

## Compilation
Scripts can be at different stages of compilation, but the process generally consists of a linear series of steps: tokenizing, parsing, analyzing, and finally compiling.

The main class of the GDScript module is the [`GDScript`](gdscript.h) class, which represents a class defined in GDScript. Each `.gd` file is called a _class file_ because it implicitly defines a class in GDScript, and thus results in an associated `GDScript` object. However, GDScript classes may define [_inner classes_](https://docs.godotengine.org/en/stable/tutorials/scripting/gdscript/gdscript_basics.html#inner-classes), and those are also represented by further `GDScript` objects, even though they are not in files of their own.

The `GDScript` class contains all the information related to the corresponding GDScript class: its name and path, its members like variables, functions, symbols, signals, implicit methods like initializers, etc. This is the main class that the compilation step deals with.

> A secondary class is `GDScriptInstance`, defined in the same file, containing _runtime_ information for an instance of a `GDScript`, and is more related to the execution of a script by the virtual machine.

Author: I opted away from mentioning `PropertyInfo/MethodInfo`, as I think that's getting a little into the weeds of the analyzer, but added a section on how GDScript interfaces with Godot, using the classes you mention!

### Loading source code

> This mostly happens by calling `GDScript::load_source_code()` on a `GDScript` object. Parsing only requires a `String`, so it is entirely possible to parse a script without a `GDScript` object!

Author: I think this is a good idea! Though we need to remember to change it if the feature (which is hotly requested...) ever does get implemented :)

See https://github.com/godotengine/godot-proposals/issues/4220 and recent discussions in https://github.com/godotengine/godot-proposals/issues/6031

> ### Tokenizing (see [`gdscript_tokenizer.h/cpp`](gdscript_tokenizer.h))

Reviewer 4: ```suggestion
### Tokenizing (see [`GDScriptTokenizer`](gdscript_tokenizer.h))
```

> Tokenizing is the process of converting the source code `String` into a sequence of tokens, which represent language constructs, identifiers, literals, etc. It happens as the first step of the parsing process, and so isn't really found independently of that in the codebase, save very rare exceptions.

Author: I've actually very rarely touched the tokenizer and parser, so I trust your expertise here. What I noticed is that references to the tokenizer only ever shows up in the 1) parser, 2) gdscript editor, 3) highlighter, and 4) test framework. I think you're right though, the highlighter doesn't really seem to use it a lot, just references some tokens.

> Tokenizing is the process of converting the source code `String` into a sequence of tokens, which represent language constructs, identifiers, literals, etc. It happens as the first step of the parsing process, and so isn't really found independently of that in the codebase, save very rare exceptions.

Author: I rewrote this a little bit, because the exception was meant to say that we don't see it independent of the parser, except in e.g. the code editor (which is one of the rare exceptions) :)

> ### Parsing (see [`gdscript_parser.h/cpp`](gdscript_parser.h))

Reviewer 4: ```suggestion
### Parsing (see [`GDScriptParser`](gdscript_parser.h))
```

The parser takes a sequence of tokens and builds the abstract syntax tree (AST) of the GDScript program. This is what finds any syntax errors in a GDScript program. The AST is used in the analyzing and compilation steps, and the source code `String` and sequence of tokens are discarded. 

> The parser class also defines all the possible nodes of the syntax tree, e.g. `GDScriptParser::ClassNode` for a node representing a GDScript class. It further defines the typechecking datastructures like `GDScriptParser::Datatype`.

Reviewer 4: It may be worth mentioning the parser limitations that we [deliberately added](https://godotengine.org/article/gdscript-progress-report-writing-new-parser/#less-lookahead). The parser makes **one** pass: unlike the analyzer, some nodes have not yet been parsed. At any given time, only 2 tokens are available to the parser: the previous and the current. Understanding these facts has helped me in several PRs when I was trying to do something that was impossible.

Author: Yes!!!! I totally forgot that that was a super intentional design decision that is _absolutely_ worth mentioning here!

> ### Analysis and typechecking (see [`gdscript_analyzer.h/cpp`](gdscript_analyzer.h))

Reviewer 4: ```suggestion
### Analysis and typechecking (see [`GDScriptAnalyzer`](gdscript_analyzer.h))
```

In practice, full scripts are simply scripts where `GDScript::reload()` has been called. This critical function is the primary way in which scripts get compiled in Godot, and essentially does all the compilation steps covered so far in order. Whenever a script is loaded, or updated and reloaded in Godot, it will end up going through `GDScript::reload()`, except in very rare circumstances like the test runner. It is an excellent place to start reading and understanding the GDScript module!

## Special types of scripts

Certain types of GDScripts behave slightly differently. For example, autoloads are loaded with `ResourceLoader::load()` during `Main::start()`, very soon after Godot is launched. Many systems aren't initialized at that time, so error reporting is often significantly reduced and may not even show up in the editor.

Tool scripts, declared with the `@tool` annotation on a GDScript file, run in the editor itself as opposed to just when the game is launched. This leads to a significant increase in complexity, as many things that can be changed in the editor may affect a currently executing tool script.

## Other

> There are many other classes in the GDScript module. Here is a brief overview of some of them:

Author: Yes, I had forgotten about these, but I agree: the readme isn't meant to be fully comprehensive, but merely a starting point for folks wanting to get into basic GDScript work. If they want to get into those advanced topics, they will already be familiar enough with this that they can independently go figure those out :)
